gateway-ha/src/test/java/io/trino/gateway/ha/router/TestRoutingGroupSelector.java

```diff
@@ -111,7 +111,6 @@ void testByRoutingRulesEngineNoMatch(String rulesConfigPath)
         assertNull(routingGroupSelector.findRoutingGroup(mockRequest));
     }
 
-    //Todo: The functionality of reading the file before every request needs to be smarter
```

Comment:
Maybe we need to find a way to NOT read the file before every request and instead have some mechanism to reload it into memory when it gets changed .. or maybe the change triggers the reload

Reply:
I removed this comment because it isn't accurate. Tl;dr, a change does trigger a reload. A little history - 

The Presto Gateway reloaded the rules on every new request. Besides being inefficient, this was actually broken, because the `snakeyaml` parser is not thread safe. Routing would randomly fail as request concurrency increased.

I submitted a fix last year that added locking to prevent multiple threads from attempting to parse the rules file simultaneously. Making every request wait for a read lock could limit performance, so the timestamp check was added to check if the file itself was updated. If the file hasn't changed, the file doesn't need to be parsed and the current rules can be used. This is why the file timestamp needs to increment for this test to work.

We could add an API method to reload rules - but this would force every Gateway user to modify their operations, so I think the decision should be carefully considered. I also don't think this is a real world problem, CICD is the only place where you would modify the rules file < 1 ms after loading it.

Reply:
Thanks for the background .. and agreed.